cc/README.md

```diff
@@ -9,115 +10,188 @@ build system. We have experienced build issues with the latest version of Bazel,
-appropriate, however a lot of the implementation details are different for
-performance reasons. In particular, the logic that handles the board state
-strives for a small memory footprint and eschews data structures more
-sophisticated than a simple array in an effort to minimize memory allocations
-and maximize cache locality of data. In doing so, the implementation is kept
-fairly simple (there is no need for a LibertyTracker) and at the time of writing
-performance of the Position code was more than 450x that of its Python
-counterpart.
+Loads all SGF files found in the given `sgf_dir` and tries to predict the move
+made at each position in each game.
 
-## Inference engines
+```shell
+bazel build -c opt cc:puzzle
+bazel-bin/cc/puzzle \
+  --model=saved_models/000990-cormorant.pb \
+  --num_readouts=160 \
+  --sgf_dir=puzzle_sgf
+```
 
-C++ Minigo currently supports three separate engines for performing inference:
+## Running the unit tests
 
- - tf: peforms inference using the TensorFlow libraries built by
-   `cc/configure_tensorflow.sh`.
- - lite: performs inference using TensorFlow Lite, which runs in software on
-   the CPU.
+Minigo's C++ unit tests operate on both 9x9 and 19x19, and some tests are only
+enabled for a particular board size. Consequently, you must run the tests
+multiple times:
 
-The Compilation and linking of these engines into the `//cc:main` binary is
-controlled by the Bazel defines `--define=tf=<0,1>` and `--define=lite=<0,1>`.
+```shell
+bazel test --define=board_size=9 cc/...  &&  bazel test cc/...
+```
 
-The choice of which engine to use is controlled by the command line argument
-`--engine=<tf,lite>`.
+## Running with AddressSanitizer
 
-## TensorFlow Lite
+Bazel supports building with AddressSanitizer to check for C++ memory errors:
 
-Minigo supports Tensorflow Lite as an inference engine.
-Build `//cc:main` with `--define=lite=1` and run with `--engine=lite`.
+```shell
+bazel build cc:selfplay \
+  --copt=-fsanitize=address \
+  --linkopt=-fsanitize=address \
+  --copt=-fno-omit-frame-pointer \
+  --copt=-O1
+```
+
+## Inference engines
+
+C++ Minigo currently supports multiple separate engines for performing
+inference. Which engines are compiled into the C++ binaries are controlled by
+passing Bazel `--define` arguments at compile time. The C++ binaries can choose
+to use any of the compiled engines at runtime using the `--engine` command line
+argument:
+
+ - **tf**: peforms inference using the TensorFlow libraries built by
+   `cc/configure_tensorflow.sh`. Compiled & used as the inference by default,
+   disable the engine with `--define=tf=0`.
+ - **lite**: performs inference using TensorFlow Lite, which runs in software on
+   the CPU.
+   Compile by passing `--define=lite=1` to `bazel build`.
+   Use by passing `--engine=lite` when you run your compiled binary.
+ - **tpu**: perform inference on a Cloud TPU. Your code must be compiled and run
```

Comment:
the compilation has to happen on a ctpu too?

Reply:
I've removed that mention, though I don't think we've ever tried compiling on a different machine. The model definitely has to be frozen on a Cloud TPU machine.